Create operators.py file

szymag opened the pull request:
Create classes `Gradient`, `Curl`, `Divergence` where expression
is not evaluated. 

Functions `curl`, `divergence`, `gradient` now
are localized in `operators.py` file.

Argument `coord_sys` in gradient, curl, divergence, have default value `None`.
This parameter is obtained from vector or scalar itself in function `_get_coord_sys_from_expr`

szymag committed:
Create operators.py file.

Create classes Gradient, Curl, Divergence where expresion
is not evaluated. Functions curl, divergence, gradient now
create object from proper class and evaluate expression.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,154 @@
+    """
+
+    def __new__(cls, scalar, coord_sys):
+        return Del(coord_sys).gradient(scalar)
```
This should not calculate the result, write something like:
```python
obj = Expr.__new__(...)
return obj
```
Same for the other classes.

Move calculation to `def doit(self, ...)` method.

szymag replied:
When I move calculation to doit, shoud I add `.doit()`?

Upabjojr replied:
Not in the constructor.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,154 @@
+from sympy.vector.deloperator import Del
+
+
+class Gradient(Basic):
```
Inherits `Expr`

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,154 @@
+        return Del(coord_sys).cross(vect)
+
+
+def gradient(scalar, coord_sys):
```
Default argument `coord_sys=None`, then put an `if coord_sys == None:` in the function code. Same for the other functions.

Upabjojr replied:
Try `v2.atoms(CoordSysCartesian)` or something similar. The information is always contained in the expression, you just need to find it.

szymag replied:
It works! Thank you 🙂

Upabjojr commented:
Some tests for the new additions?

Upabjojr commented:
Tests should just check that `.doit()` works and that expressions are not evaluated if no `.doit()` is called.

szymag commented:
> Tests should just check that .doit() works and that expressions are not evaluated if no .doit() is called.
OK, I will do that

szymag committed:
Apply sugestion from comments.

Add tests, change inheritance, argument
coord_sys have default value None.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,184 @@
+
+    """
+
+    def __new__(cls, scalar, coord_sys):
```
Let's remove the `coord_sys` dependency, let's define the operators as close as possible to how they are defined in the math books.

szymag replied:
OK. I will remove this parameter and obtain it in similar manner as in funtions

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,184 @@
+    """
+
+    if coord_sys is None:
+        coord_sys = list(scalar.atoms(CoordSysCartesian))[0]
```
Put this in a private function, like `_get_coord_sys_from_expr(...)`, this code is shared by all operators.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,184 @@
+        coord_sys = list(scalar.atoms(CoordSysCartesian))[0]
+        return Gradient(scalar, coord_sys).doit()
+    else:
+        return Gradient(scalar, coord_sys).doit()
```
Repetition of code.

szymag committed:
Remove coord_sys parameter from constructor.

Add method _get_coord_sys_from_expr.

szymag commented:
I applied your suggestion. I think that now it has more sense. 
Now, functions gradient, curl, divergence can return value in two ways, as before and using new classes.
I think that it's worthwhile to add description about it.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,187 @@
+        obj._scalar = scalar
+        return obj
+
+    def _get_coord_sys_from_expr(self, scalar):
```
I meant this one as a global function, not a class method. Let's avoid repeating the code. Also, this could be used in the operator functions.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,187 @@
+        return vect._sys
+
+    def doit(self):
+        return Del(self._get_coord_sys_from_expr(self._vect)).dot(self._vect).doit()
```
What about putting this code in the functions? I think it's clearer if `doit( )` just calls the respective function (in this case `divergence`).

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,187 @@
+    """
+
+    if coord_sys is None:
+        return Divergence(vect).doit()
```
I would just get the coordinate system like `coord_sys = _get_coord_system_from_expr(vect)`. I am insisting on doing so because I'd like to put more code here in the future.

szymag replied:
Do you want create these functions independently on already created classes?

Upabjojr commented on sympy/vector/tests/test_operators.py:
```diff
@@ -0,0 +1,29 @@
+    assert Gradient(s1) == Gradient(R.x*R.y*R.z)
+    assert Gradient(s2) == Gradient(R.x + 3*R.y**2)
+    assert Gradient(s1).doit() == R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
+    assert Gradient(s2).doit() == R.i + 6*R.y*R.j
```
OK, that's what I meant.

Upabjojr commented:
> I think that it's worthwhile to add description about it.

Let's work on the code first. Once we're happy about the code, we can write other stuff like documentation.

szymag commented:
> I meant this one as a global function, not a class method. Let's avoid repeating the code. 

What do you think about creating class `Operator` and then three inherited classes?

szymag committed:
Extract _get_coord_sys_from_expr as global function.

Simplify in functions.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,182 @@
+
+    """
+
+    return _get_coord_sys_from_expr(scalar).delop(scalar).doit()
```
Add this warning (we need to maintain compatibility):

```python
from sympy.utilities.exceptions import SymPyDeprecationWarning

if coord_sys is not None:
    SymPyDeprecationWarning(
        feature ="coord_sys parameter",
        useinstead ="do not use it",
        deprecated_since_version ="1.1"
    ).warn()
```

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,182 @@
+        return Del(_get_coord_sys_from_expr(self._vect)).cross(self._vect)
+
+    def doit(self):
+        return self.curl().doit()
```
I mean `return curl(self)`

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,182 @@
+        return Del(_get_coord_sys_from_expr(self._scalar)).gradient(self._scalar)
+
+    def doit(self):
+        return self.gradient().doit()
```
I mean `return gradient(self)`, calling the function.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,182 @@
+        obj._scalar = scalar
+        return obj
+
+    def gradient(self):
```
We already have the function in the global namespace, we don't need a repetition.

Upabjojr commented:
> What do you think about creating class Operator and then three inherited classes?

Only necessary if the subclasses inherit something. In this case they don't have anything in common except for the methods of `Expr`. I don't think we need this.

Upabjojr commented:
Good start for the first day. I will not be available tomorrow, so see you on Friday.

szymag commented:
Thank you for your attention. I will work on a documentation and proposal of introducing Lame coefficient.
See you on Friday.

szymag committed:
Small clean up in operators.py. Add warning.

_get_coord_sys_from_expr has warning of using
coord_sys parameter.

Upabjojr committed:
doit( ) with **kwargs and sympify in constructors

Upabjojr commented:
@szymag you should add some entry tests in `sympy/core/tests/test_args.py` for the newly created classes.

Upabjojr commented:
Also:

- we have now duplicate definitions of the functions (`sympy.vector.functions`), maybe the functions should be moved to that file and stay there.
- lots of tests now throw a deprecation warning, they need to be changed as no coordinate system parameter should be passed to them.

szymag commented:
> you should add some entry tests in sympy/core/tests/test_args.py for the newly created classes.

I will do that.

>  (`sympy.vector.functions`), maybe the functions should be moved to that file and stay there.

I'd like to keep functions and classes in one place. I think it has more sense. I wanted to delete rather these functions from `sympy.vector.functions`.

>lots of tests now throw a deprecation warning, they need to be changed as no coordinate system parameter should be passed to them.

Yes, I see. I will look into tests and remove `coord_sys` from parameters. 

I would like to also change one name of parameter, from `vect` to `vector`. Saving characters is negligible and full name looks much better.

Upabjojr commented:
> I wanted to delete rather these functions from sympy.vector.functions.

We cannot break the API, many people may be using the code with the functions in those positions, we don't want to cause widespread problems.

> I would like to also change one name of parameter, from vect to vector.

I changed it into `expr`, because it could in principle be any expression (think about `Divergence(Gradient(...))`.

szymag commented:
> We cannot break the API, many people may be using the code with the functions in those positions, we don't want to cause widespread problems.

Yes I know. You've told me that. You also suggest creating link in `sympy.vector.functions` to `sympy.vector.operators`. I'm working on that, but simple `from sympy.vector.operators import divergence, curl, gradient` doesn't work. I have to find solutions. 

> I changed it into expr, because it could in principle be any expression (think about `Divergence(Gradient(...))`.

It's OK, but in functions you have still `scalar` and `vect`.

szymag committed:
Remove unnecessary things.

Remove coord_sys parameter from test and doctest,
remove old function in vector/functions,
apply PEP8 in in one file with test.

szymag committed:
Add tests to core/tests/tests_args.py.

Change parameter vect to vectors in operators
functions.

szymag commented:
There should be still one warning due to `coord_sys` parameter. I must leave explicit information about coordinate system in one test:
```python
from sympy.vector.vector import Vector
from sympy.vector.coordsysrect import CoordSysCartesian

C = CoordSySCartesian('C')
curl(Vector.zero, C)
```

Vector.zero has no information about coordinate system, because it's created independently. I think it's topic for another PR.

Upabjojr commented:
We need to fix it somehow, otherwise I believe the tests will fail. Maybe a temporary `if`?

szymag commented:
> We need to fix it somehow, otherwise I believe the tests will fail. Maybe a temporary if?

I'm also thinking about it. 
For me, it seems to be quite hard problem to solve. I'd say that we should change `VectorZero` class, but it breaks API.

Upabjojr commented:
Let's have a quick edit for this, then merge. The vector module is a bit of a mess and some code cleanup is wished.

Maybe you could add
```python
if len(expr.free_symbols) == 0:
    return S.Zero
```

Upabjojr commented:
I would also like to get rid of the horrible way that `VectorAdd` and `VectorMul` and handled. In my opinion, those classes are useless.

szymag commented:
>I would also like to get rid of the horrible way that VectorAdd and VectorMul and handled. In my opinion, those classes are useless.

We can work on it in the next step. I believe it will be easier later.

Upabjojr commented:
Yes, let's focus on the main topics of the GSoC first.

szymag commented:
`S.zero` doesn't work. I get error:
```python
AttributeError: 'Zero' object has no attribute 'delop'
```
I should return coordinate system.

szymag commented:
I put it in `_get_coord_sys_from_expr` but now I think you proposed to put it in curl, yes?

szymag committed:
Add exception in operators file.

szymag commented:
I think it should pass now. I had to add another exception for `0`. We can put scalar field as int in one case, when is equal to 0.

szymag committed:
Repair test name in test_args.py

Upabjojr commented on sympy/core/tests/test_args.py:
```diff
@@ -3982,6 +3982,28 @@ def test_sympy__vector__scalar__BaseScalar():
     assert _test_args(Wigner3j(0, 0, 0, 0, 0, 0))
+
+
+def test_sympy__vector__operators_Gradient():
```
two underscores after `operators`, same for the next ones.

szymag replied:
I have already repaired this

szymag commented:
Do you like changes in operators.py? I have to add `if` statement in `_get_coord_sys_from_expr` but also in `gradient`, `curl` and `divergence`.

Upabjojr commented on sympy/vector/operators.py:
```diff
@@ -0,0 +1,202 @@
+    10*R.x*R.z*R.i + 5*R.x**2*R.k
+
+    """
+    if _get_coord_sys_from_expr(scalar, coord_sys) is False:
```
You're calling this function twice. I'd have `coord_sys = _get_coord_sys_from_expr(...)` once. Also I'd prefer a `None` instead of `False` (let's be precise with the returning types).

szymag replied:
You're right. I will update my changes.

szymag committed:
Clean up in operators.py.

szymag commented:
When we merged this PR, I'd like to remove `coord_sys` from functions in `functions.py` file, like `scalar_potential_difference(field, coord_sys, point1, point2)`.
It's easy task.

Upabjojr committed:
inherit Expr instead of Basic
